### ISSUE TITLE
Method log_prob_with_logits() for Dirichlet

### ISSUE THREAD

**Issue by akuz** [tensorflow/tensorflow#7956]
It would be useful to have a log_prob_with_logits() method for the Dirichlet distribution.

The reason being is that it is often useful to model the discrete posterior distribution in log space, which doesn't let all probabilities to go exactly to zero. Then the Dirichlet prior can be applied to the data in log space. Note, that if the posterior is converted to the normalised distribution, then some of the discrete probabilities may actually go to zero due to the rounding errors. Then the Dirichlet prior cannot be applied to this distribution because it doesn't allow zero probabilities.

The log_prob_with_logits() calculation would be very simple to implement. The log(x_i) would need to be replaced with just x_i, and an additional term with LogSumExp(x) added.

This would actually enable the possibility to use the Dirichlet prior on the discrete viariables in log space (can also be part of a neural network).

**Comment by akuz** [comment]
Namely, the formula for log_prob_with_logits(x) would be as follows (where x are logits, and L is the log of the Dirichlet normalisation constant):

![formula](https://cloud.githubusercontent.com/assets/2409854/23442260/5dc438cc-fe1f-11e6-813d-220f9be70c85.jpg)

**Comment by tatatodd** [comment]
@langmore Can you comment on this?  Thanks!

**Comment by langmore** [comment]
Good point akuz.  Rather than adding a new method, I think it would be better to allow the Dirichlet to be initialized either with `concentration` (as it currently is) OR `log_concentration`.  See e.g. the `Categorical` distribution for an example.

I'm choosing `log_concentration` rather than `logits`, since to me, `logits` are something that is intended to be fed through a softmax, and therefore they are only defined up to an additive constant.  Whereas here, the absolute value matters.

Does this sound like something you would want to do?

**Comment by akuz** [comment]
Hi Ian, I guess this could also accept log_concentration in parameter (although I don't have a use case for that). This still doesn't solve the problem I was talking about. The Dirichlet distribution is defined *over* the categorical distributions. These are the concrete distributions that can be passed as x to log_prob(x). They have to be valid distributions, I.e. sum up to one, and all numbers must be > 0 and < 1. The Dirichlet doesn't allow any of the components to be exactly 0 or 1, and will output NAN in such case. Now imagine that the input to log_prob(x) is such that x = softmax(y). This can easily result in the output like x = [0, 0, 1, ..., 0, 0], for which the Dirichlet log_prob(x) will output NAN. So what I am suggesting is to pass logits directly to log_prob() without applying softmax first. We can do it like Dirichlet.log_prob_with_logits(), which would be similar to cross_entropy_with_logits(). With cross entropy they don't apply softmax before passing to the function for exactly the same reason of potential rounding of the output of softmax to x = [0, 0, 1, ..., 0, 0]. (Aternatively, instead of making a new method, we could add a new parameter to log_prob(x, logits) such that only one of x or logits can be specified. What do you think?

**Comment by langmore** [comment]
You're right, I was mistaken.  The problem with zeros or ones in `x` needs some sort of `log_prob_with_logits` function.  I suppose there is another problem if `concentration` has any zeros in it, but that would require a new version of the Gamma function.

So the solution suggested by @tatatodd is to add a `log_prob_with_logits` function for distributions defined on the interior of the simplex, `{(x_1,...,x_K) :  x_i \in (0, 1), \sum_i x_i = 1}`.  This would be `Dirichlet`, and `Beta`.

 Note however that `Gamma.log_prob(x)` will fail for the exact same reason if `x = 0` (both compute `log(x)`).  In fact, all of our distributions defined on `(0, inf)` have this issue.  In my work, I've avoided this by careful use of priors, but before I had this down it was a pain.  It would be good if the solution worked for these distributions as well.  Or, @tatatodd do you think that `Dirichlet` is more prone to this issue?

@ebrevdo What do you think?

**Comment by akuz** [comment]
I agree about Dirichlet and Beta. Not sure about Gamma, maybe log is not the best transform in that case.

**Comment by ebrevdo** [comment]
I'd prefer a new Dirichlet distribution defined over the logits space,
rather than adding a new method to all distributions.  wdyt?

**Comment by akuz** [comment]
Actually, sounds good, abrevdo. I haven't worked with probabilistic inference much in TensorFlow yet, but I see there is a bunch of distributions already following your suggestion.

**Comment by langmore** [comment]
@ebrevdo , If we simply use the formula above then we do not have the log of a probability density over `R^k`, so we can't call it `log_prob(x)`.  To see that the above formula is not a `log_prob`, note that it doesn't go to negative infinity as `|x| \to \inf`, so the corresponding `prob(x) := exp(log_prob(x))` will not go to zero.  You can see also that the above formula is constant if all the `x_i` are equal.  The reason being that whenever all the `x_i` are equal, `Softmax(x) = [1/k,...., 1/k]`.  So the entire line `{x : x_1 = x_2 = ... = x_k}` is mapped by Softmax to a single point.

I believe @akuz wants the standard Dirichlet distribution over `k-simplex` of probabilities, but wants to evaluate it on an intermediate state of data (before the probabilities are probabilities).

Note @akuz that since this doesn't give you a valid prior on logits, it may be better to create the distribution over logits that corresponds to a specified Dirichlet distribution over probabilities.  After all, your network produces logits.  This can be done as follows:

```python
ds = tf.contrib.distributions
bijectors = tf.contrib.distributions.bijectors  # you'll have to search for this...depending on version.

# The dirichlet is a distribution over the k-1 dimensional k-simplex
dirichlet = ds.Dirichlet(concentration=[1., 2., 3.])
softmax_centered = bijectors.SoftmaxCentered(event_ndims=1)

# dist_over_logits is a distribution over R^{k-1}
dist_over_logits = ds.TransformedDistribution(
    dirichlet,
    bijector=bijectors.Invert(softmax_centered))

logits = [-1., 1.]
probs = softmax_centered.forward(logits)

# These two are equal
probs
==> [ 0.09003057  0.66524088  0.24472845]
tf.nn.softmax(logits + [0.])  # Note implicit appending of [0]
==> [ 0.09003057  0.66524088  0.24472845]

# These two are not equal
dirichlet.log_prob(probs)
==> 0.871526
dist_over_logits.log_prob(logits)
==> -3.35129
```

The last two lines above are not equal because the Dirichlet is a distribution over the (k-1 dimensional) k-simplex, which is very different (in particular, it is much smaller) than R^{k-1} in which logits live.  The probability density over R^{k-1} must be much smaller at the point `logits` that corresponds to `probs`

**Comment by langmore** [comment]
....in the above example, `dirichlet` is the distribution for a random variable `Z = Dirichlet([1, 2, 3])`, and `dist_over_logits` is the distribution of the random variable `X = InverseSoftmaxCentered(Z)`, i.e. `Z = SoftmaxCentered(X)`.  In other words, `Z` is the correct distribution for your logits.

**Comment by akuz** [comment]
@langmore you're right that the above formula doesn't define a proper density on the k-D space of logits, due to the reasons you mentioned, so it would not make sense to call it log_prob(x)... The transformed distribution trick you suggested looks great, but I just looked through the documentation, and it seems SoftmaxCentered only supports low dimensionality inputs. What I have is a 4D tensor, the last dimension of which would be the logits, so I'm afraid the transformed distribution might not work in this case? Also would that be efficient?

On the other side, if you ask yourself the following question: "Does the above formula calculate log probability of softmax(x) under the specified Dirichlet prior?" - the answer is yes. So, it does not provide the density in the space of logits, but rather in the space of discrete distributions resulting from softmax(x). So, all I'm trying to do is: 1) y = softmax(x), 2) Dirichlet(concentrations).log_prob(y). But because softmax has a tendency to collapse to [0, 0, ..., 1, 0] due to rounding, the Dirichlet log_prob fails to work properly. The above formula would provide a shortcut to compute steps 1) and 2) in one go, without actually computing softmax. Yes, it would not define a valid density on x, and therefore maybe should not be implemented as a Distribution. Maybe it would be better if implemented as some sort of function on tensors? dirichlet_log_prior_on_softmax_with_logits(concentrations, logits)? Well, maybe a better name would be needed :)

So, it short, I'm not looking to define any new types of Distributions, but an efficient way to compute steps 1) and 2) above in one go, without actually computing the softmax in between. Similar concept to cross_entropy_with_logits().

**Comment by akuz** [comment]
In fact, if we ignore the normalisation constant for now, the above formula can be implemented using the convolution (where concentrations would be used as weights for the last dimension) for the first part, and reduce_logsumexp() for the second part (you would also need to compute and multiply it by the negative of the sum of concentrations). But this implementation has two drawbacks 1) you need to configure the convolution in a special way, and it is only implemented for some specific shapes of tensors, whereas here we always need to convolve along the last dimension only, and 2) it produces two intermediate tensors, which are the summed up. Therefore, a specific implementation would be much more efficient.

**Comment by langmore** [comment]
@akuz , I agree that your original suggestion does provide an alternative method to compute the log_prob on the space of probabilities by intercepting an intermediate step of computation of an input.

So.... we can't call this `log_prob` since it isn't one.  akuz's original suggestion of `log_prob_with_logits(x)` would work (although I think the `alpha` in the formula should be `alpha - 1`).  @ebrevdo I think the options are:

1. Do nothing, let users build their own (`Dirichlet._log_unnormalized_prob` will help BTW)
1. Write a helper function, e.g. `dirichlet_log_prob_with_logits(dirichlet_dist, logits)`.
1. Create a `Dirichlet.log_prob_with_logits(logits)` method.

BTW, the dimensionality restriction on SoftmaxCentered is that it works for batch scalars and vectors (event_ndims = 0 or 1), not batch matrices (event_ndims = 2).  So it would work well for a Tensor of shape [2, 3, 4, 5], representing a shape [2, 3, 4] batch of length [5] logits.  But, it seems this is not what you need.

**Comment by akuz** [comment]
@langmore , I think that option 3 from your list of suggestions would be most convenient to use in practice. Sorry about the mistake missing -1 in the formula. The computation of the "L" part in the formula should be possible to do easily too (I've done it with lngamma function in C++ previously). I think I would not be the only one to need this, hopefully.

**Comment by akuz** [comment]
@langmore I was looking more into prior distributions defined on the R+, such as Gamma, InverseGamma, etc. I think these would benefit from something like log_prob_with_softplus(). I've made a simple model of clustering using expectation maximisation. In it, the latent variables (cluster assignments) and the parameters of the clusters (cluster probs, means, stdevs) are estimated using a single optimizer loop (all together). If I put the stdevs directly as the variables to optimize, this leads to problems, because optimizers don't allow specifying the bounds. So what I ended up doing is:

stdevs_raw = tf.Variable(...)

stdevs = tf.softplus(stdevs_raw) - this makes the stdevs always positive

stdevs_prior = tf.constrib.distributions.InverseGamma(100, 10).log_prob(tf.pow(stdevs, 2))

However, this still has a problem if one of the stdevs collapses to zero (which is possible if stdevs_raw becomes a large negative number).

Therefore, it would be useful to have something like InverseGamma.log_prob_with_softmax(x) which would compute the log probability for softmax(x) without actually computing the softmax.

Maybe this can be done with some variable transformations? I am not sure

**Comment by langmore** [comment]
Question:  Why did you feed `stdevs ** 2` to `log_prob` rather than `stdevs`?

I see this as a good features request.  We have to figure out if there is a solution that will work across different distributions.  Building upon your previous suggestion, what about `log_prob_pre_X`, where `X` is `softmax`, `softplus`, `exp`, etc...  ?

I don't think this will be resolved immediately.  So for the moment I recommend writing helper functions.  This will actually be better since you can see what works and give us feedback.  If you want, you could add a helper function to the same file that contains the class and submit a pull request.  E.g. add `log_prob_pre_softmax(dist, logits)` to `dirichlet.py`, along with some tests.  For the moment, keep it hidden, i.e. don't add it to the `__all__` list at the top of the file.  If this helper function works well, we can either un-hide it, or add it as a method.

Also, and I realize every use case is different, let me tell you how I've been able to solve this issue.  So far, I've been able to avoid the collapse to zero (with 32 bit tensors) by using `AdamOptimizer` with a small step size of around 0.01 and making sure that the prior's log-prob was included in my loss (you're already doing this I assume, but just in case...).  So in your case above I make sure my loss is `-1 * log_likelihood - 1 * stdevs_prior`.  In theory, the penalty from the prior should prevent collapse to zero, but in practice of course many crazy things can happen, especially when you have hidden layers.

**Comment by akuz** [comment]
@langmore , I fed stdevs ** 2 to InverseGamma, because inverse gamma is a conjugate prior for the variance of Normal with known mean. The conjugacy doesn't play much role here, but there is another benefit. The conjugate prior hyperparameters have interpretation that is easy for me to use to set a specific prior, see here: https://en.m.wikipedia.org/wiki/Conjugate_prior#Continuous_distributions (see "Interpretation of hyperparameters" column).

I agree that would be a useful feature (partially because I've created this request :)), because of the collapse of the posteriors to zeroes, which renders many useful priors invalid, if not calculated in log space directly. As you say, it's not clear what would be the best way to implement it in conjunction with what already exists. When I have time, I will try to follow your suggestion, but I have to spend most of my time on my full time job. Hopefully I can find some time on lunch breaks / evenings soon.

**Comment by akuz** [comment]
@ebrevdo - just saw the new paper that came out "Deep Probabilistic Programming" - how do you deal with the problem described in this ticket within the Edward lib?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by langmore** [comment]
I'll bring this up to the team.  I think the way to think of this is that we would offer a version of `Dirichlet` (and others) that has been "pre-transformed" (in a stable manner) to unconstrained space using some `Bijector`.  E.g. `DiricheletUnconstrained` would be a stable equivalent to

```python
tfd = tf.contrib.distributions
tfb = tfd.bijectors
dist = tfd.Dirchlet([0.5, 0.5])
dist_uc = tfd.TransformedDistribution(
    dist,
    bijector=tfb.Invert(tfb.SoftmaxCentered()))
```

**Comment by dustinvtran** [comment]
I like your solution @langmore to use transformed distribution. Among other things, a canned implementation could be useful for Dirichlet priors on Categorical/Multinomial logits.  

How do you propose handling the fact that the desired function doesn't normalize? If I understand the discussion correctly, `dist_uc.log_prob` does _not_ return the same as "`log_prob_pre_softmax`"?

> @akuz: @ebrevdo - just saw the new paper that came out "Deep Probabilistic Programming" - how do you deal with the problem described in this ticket within the Edward lib?

This is actually a problem in Edward's mixture models. We just work in the `probs` space and it can be unstable.

Examples:
+ [`examples/stochastic_block_model.py`](https://github.com/blei-lab/edward/blob/93b6e31dce17c2ad3dca46b80a825819f02da5fe/examples/stochastic_block_model.py)
+ [`examples/mixture_gaussian_gibbs.py`](https://github.com/blei-lab/edward/blob/master/examples/mixture_gaussian_gibbs.py)

**Comment by langmore** [comment]
Note that I'm not suggesting we use transformed distribution (since what I wrote above would be no more stable than `Dirichlet`.  I'm suggesting we pre-bake an equivalent distribution.

I don't understand the normalization issue.  Perhaps tell me what the args are above to `log_prob` and `log_prob_pre_softmax`?

**Comment by dustinvtran** [comment]
Oh I see. I'm only for a canned distribution over logits space;  so I still agree :)

I don't understand the normalization issue either. The above conversation is rather involved—something about `log_prob_with_logits(x)` in the latex formula not defining a proper density? Maybe you can summarize how that relates to your "implement a canned distribution" proposal?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dynamicwebpaige** [comment]
Given that [this endpoint](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Dirichlet) has moved to TensorFlow Probability, it would make sense to transfer the feature request. 👍 

cc: @jvdillon

**Comment by mohantym** [comment]
Hi @akuz ! Could you transfer this feature request to [TF probablity ](https://github.com/tensorflow/probability)repo as per above [comment](https://github.com/tensorflow/tensorflow/issues/7956#issuecomment-546747505).  Thank you!
